### ISSUE TITLE
[QUERY] Azure SDK for Java: Should Key Vault Client swallow exceptions?

### ISSUE THREAD

**Issue by mattsmall1972** [Azure/azure-sdk-for-java#36247]
**Query/Question**
When the Java Key Vault client fails while sending a HTTP request, it captures the exception, writes a log about the failure, and continues execution.  Is it proper design to swallow the exception without giving the calling code ability to handle the exception?

```
 try (CloseableHttpClient client = buildClient()) {
            HttpGet httpGet = new HttpGet(url);
            if (headers != null) {
                headers.forEach(httpGet::addHeader);
            }
            httpGet.addHeader(USER_AGENT_KEY, USER_AGENT_VALUE);
            result = client.execute(httpGet, createResponseHandler());
        } catch (IOException ioe) {
            LOGGER.log(WARNING, "Unable to finish the http get request.", ioe);
        }

```

***Why is this not a Bug or a feature Request?***
It is unclear if this is by-design or a bug.

**Setup (please complete the following information if applicable):**
 - Library/Libraries: Azure SDK for Java / Azure Key Vault
 
[https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/HttpUtil.java#L62](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/HttpUtil.java#L62)

 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X ] Query Added
- [X ] Setup information Added

**Comment by joshfree** [comment]
@vcolin7 could you please follow up with @mattsmall1972?

**Comment by vcolin7** [comment]
@joshfree This code is a part of the Key Vault JCA library, owned by @stliu's team. I think they can handle it more appropriately.

**Comment by saragluna** [comment]
@mattsmall1972, sorry for the late reply. This is by design, and the goal is to unblock the JCA to find any other provider when the Key Vault provider is misconfigured or the certificate/key can't be found in the Key Vault. The thing is even if the exception is thrown from the Key Vault JCA provider, there's not much the client code can do to handle the exception, please share your thoughts on this.

**Comment by mattsmall1972** [comment]
> @mattsmall1972, sorry for the late reply. This is by design, and the goal is to unblock the JCA to find any other provider when the Key Vault provider is misconfigured or the certificate/key can't be found in the Key Vault. The thing is even if the exception is thrown from the Key Vault JCA provider, there's not much the client code can do to handle the exception, please share your thoughts on this.

The exception is also a warning about a problem.  Eating exceptions should be done only with proper knowledge of the exception; I don’t think that most developers recognize that this swallowed exception would be emitting a log, whereas a crashed process would definitely show up.

This same scenario was seen in earlier versions of ASP.NET.  Background threads would silently swallow exceptions so as to not crash the entire process.  This resulted in corrupted data and devs were confounded on how to troubleshoot the problem.  A crash dump was the only way to figure it out. 

 In my opinion, for this issue, only esoteric knowledge of the library would lead to the discovery of the problem.

https://stackoverflow.com/questions/2655945/asp-net-background-threads-exception-handling

**Comment by saragluna** [comment]
Yes, I agree with you on that. Swallowing the exception in the KV provider only defers it to some other component in JCA or JSSE, when the code tries to find a key/cert/some other algorithms but fails to. So users need to know the lib and JCA very well to identify the issue. 

And yes, indeed the exception is a warning about a problem. So from this perspective, we can throw the exception and let the application fail fast.
